File: lib/models/product.dart
import 'dart:convert';

import 'package:amazon_clone/models/rating.dart';

class Product {
  final String name;
  final String description;
  final double quantity;
  final List<String> images;
  final String category;
  final double price;
  final String? id;
  final List<Rating>? rating;

  Product(
      {this.id,
      required this.name,
      required this.description,
      required this.quantity,
      required this.images,
      required this.category,
      required this.price,
      this.rating});

  // convert in map
  Map<String, dynamic> toMap() {
    // print('i am in to map()');

    return {
      'product_name': name,
      'product_description': description,
      'product_quantity': quantity,
      'product_images': images,
      'product_category': category,
      'product_price': price,
      '_id': id,
      'rating': rating
    };
  }

  //from map to object
  factory Product.fromMap(Map<String, dynamic> map) {
    return Product(
        description: map['product_description'] ?? '',
        images: List<String>.from(map['product_images']),
        name: map['product_name'] ?? '',
        price: map['product_price']?.toDouble() ?? 0.0,
        quantity: map['product_quantity']?.toDouble() ?? 0.0,
        category: map['product_category'] ?? '',
        id: map['_id'],
        rating: map['ratings'] != null
            ? List<Rating>.from(
                map['ratings']?.map(
                  (x) => Rating.fromMap(x),
                ),
              )
            : null);
  }

  // from object to json
  String toJson() {
    // print('i am in tojson');
    return json.encode(toMap());
  }

// from json to map to object
  factory Product.fromJson(String source) {
    return Product.fromMap(json.decode(source));
  }

  // rating
}

File: server/models/product.js
const mongoose = require('mongoose');
const ratingSchema = require('./rating.js');





const productSchema = new mongoose.Schema({
    product_name:{
        require:true,
        type:String,
        trim:true,
    },

    product_category:{
        require:true,
        type:String,
        trim: true,
        },
    

        
    product_price:{
    required:true,
             
    type: Number,

        },

        product_quantity:{
            required:true ,
            type: Number,
            
        },

        product_images:[{
         
            required:true ,
            type: String,
            
        }],

        product_description:{
            required: true,
            type: String,
            trim: true,
        },


      ratings:[ratingSchema],

    });

const Product = mongoose.model("Product",productSchema);

module.exports ={Product,productSchema};